### PR TITLE
Rollback/Offloading Improvements

### PR DESCRIPTION
**Pull request by HardeepAsrani**
### All Submissions:

* [x] Have you followed the [Optimole Contributing guideline](https://github.com/Codeinwp/optimole-wp/blob/master/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change? 
 

### Changes proposed in this Pull Request:
This PR aims to improve the offloading/rollback process in Optimole. Based on tests run on different servers, the offloading is about 60% faster on average based on tests done on websites with 300-500 posts (images being about 3 times the posts).

I tried to also run tests on sites to compare with 1000+ posts, but it took pretty long to run head-to-head tests between the older method & newer method. Overall, things were progressing faster with the PR compared to older code.

Also includes:

- Improvements to how we handle single/selected image uploads and make it quicker instead of taking as much time as offloading all images.
- The process not runs using Cron so the user can leave the page, which makes it easier for bigger websites with thousands of posts to leave the page.
- It also has a Stop button if the user stops it.
- It also removes the estimated time, as it was not even close to being accurate.
- Take all the logic to server side to avoid thing being terribly slower if user's internet connection is weak.
- When process is ongoing and user is on the Optimole Dashboard, we send a request each 10 seconds to make sure Cron is triggered + also to update the progress bar.

I had also tried to run few consecutive cron jobs but then WP kept missing crons as some processes were faster than other, and number of cron jobs to run kept growing.

Closes https://github.com/Codeinwp/optimole-service/issues/851.

### How to test the changes in this Pull Request:

1. Make sure offloading and rollback work properly and are significantly quicker than previously.
2. Make sure single image offloading/rollback also works properly and don't take more than a couple of minutes max.
3. Make sure the website doesn't become too slow and non-accessible.
4. Make sure you can leave the page and come back to it, and it should still show the offloading/rollback ongoing.
5. It is a bit complex so would be great if you can enable debugging and keep a log of your debug.log file, and post it while reporting issues.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by HardeepAsrani** [comment]
@selul Yes, we don’t go through all the posts now but only the posts where image exists.

**Review by abaicus — APPROVED** [review]
@HardeepAsrani, except for two small things I mentioned below, things look good 👍🏻

File: assets/src/parts/connected/settings/index.js [hunk 1]
```diff
@@ -34,8 +34,7 @@ const Settings = ({
 		const {
 			getSiteSettings,
 			getQueryArgs,
-			isLoading,
-			extraVisits
```
[the remark this answers is not in the document]
Comment (HardeepAsrani) [reply assets/src/parts/connected/settings/index.js:38]: @abaicus It wasn't used in my branch; it seems like that changed after making the PR.

[the remark this answers is not in the document]
Comment (HardeepAsrani) [reply assets/src/parts/connected/settings/index.js]: @abaicus Looking at it again, it's not used anywhere. extraVisits isn't part of `select( 'optimole' )` but being defined at `extraVisits: siteSettings['banner_frontend'],`

[the remark this answers is not in the document]
Comment (abaicus) [reply assets/src/parts/connected/settings/index.js]: Indeed! Sorry for that! 🤪

**Review by abaicus — APPROVED** [review]

**Review by GrigoreMihai — CHANGES_REQUESTED** [review]
I also have a few questions about the workflow.

When you do not request specific images to be updated the flow is as before ? ie. go through all the posts and get the images that we find in those posts and at the end go through all the images that are not attached to any posts? 

I'm a bit unsure if the current WP CLI command we have for this is still working with the changes, maybe it is a good idea to add testing instructions for it ?

**Comment by HardeepAsrani** [comment]
@GrigoreMihai Yes, that db query is only when you request specific images, hence won’t affect the CLI command.

**Review by GrigoreMihai — APPROVED** [review]

**Comment by irinelenache** [comment]
@HardeepAsrani Tested and the performance is improved 🚀 , found some small issues:

- [ ] If i start the offload/rollback process and leave the page or refresh it, the progress bar will start from 0 again and that can be a bit misleading https://vertis.d.pr/v/2UGdXA
- [ ] Sometimes the rollback process doesn't work for all images. On an instance with 300 images, there were times when 30-40 images were skipped. As replicating steps, i started the rollback process and then browsed the dashboard, also created and edited a few posts (all of these while the rollback was running)

You can check the issues here:
```
 Admin area URL: https://liftedpan.s2-tastewp.com/wp-admin 
 Username: irinel 
 Password: kqaFVnOPRZI 
```

Note: No errors/warnings were thrown in the debug log 👍

**Comment by HardeepAsrani** [comment]
>  If i start the offload/rollback process and leave the page or refresh it, the progress bar will start from 0 again and that can be a bit misleading https://vertis.d.pr/v/2UGdXA

Yea, it's not the best UI but it's intended for now. It shows you process based on the remaining images, not the previous total.

**Comment by HardeepAsrani** [comment]
@irinelenache For the second issue:

- Were those new images added while offloading/rollback was ongoing? If yes, then that's expected.
- Did you try to reproduce the issue with an older build to confirm the issue already didn't exist?

I tried to offload and rollback and all but one seems to offload properly, and that one also isn't giving any errors so might not be related to the changes made.

**Comment by irinelenache** [comment]
@HardeepAsrani 

> Were those new images added while offloading/rollback was ongoing? If yes, then that's expected.

Two/three new images were added, but the rollback didn't work for 30 images

> Did you try to reproduce the issue with an older build to confirm the issue already didn't exist?

I didn't try with an older build because you couldn't leave the offload/rollback page before, without stopping the process

**Comment by HardeepAsrani** [comment]
@irinelenache You can always just stop the process or take it back all offloaded/rollbacked and start again. If the same images aren't being handled in the old build then the issue isn't with the PR. Also, maybe not create new posts while testing as we don't account for that but the images that we queue when the process starts.

**Comment by irinelenache** [comment]
@HardeepAsrani I'll do some more testing and will inform you here if i pinpoint the issue 🙏

**Comment by HardeepAsrani** [comment]
@irinelenache That will be great and specially if you could do it on instawp if possible as there I can easily access the debug log or file manager.

**Comment by irinelenache** [comment]
@HardeepAsrani I tried to test again both on TasteWP and InstaWP (on InstaWP the server is slower) and even without leaving the offload/rollback page, some images at the end of the library were skipped (last 20-30 images from about 300 images). With the live Optimole version, all images are offloaded/rollbacked

You can try on these instances:

```
Admin area URL: https://liftedpan.s2-tastewp.com/wp-admin 
Username: irinel 
Password: kqaFVnOPRZI 

https://improve-offload.vertisite.cloud/wp-admin
vohixewome8651
V2NWA0LGsJpOZRezSI58
```
Try to offload/rollback all images, go to Library and check if the process worked for the last images. If everything's fine for you, we can move this to **Ready to merge** and investigate it later

**Comment by HardeepAsrani** [comment]
@irinelenache I ran to do it few times and it seems to work fine. I made a small change which was sometimes stopping the process early but it's working fine on the instawp instance where I tested few times. The only image not being uploaded is one that has the extension type of .bmp

**Pull request merged**